include/unifex/detail/any_heap_allocated_storage.hpp

```diff
@@ -0,0 +1,161 @@
+
+namespace unifex
+{
+  namespace detail
+  {
+    template <typename T, typename Allocator>
+    class _any_heap_allocated_storage {
+    private:
+      struct state;
+
+      using allocator_type = typename std::allocator_traits<
+          Allocator>::template rebind_alloc<state>;
+      using allocator_traits = std::allocator_traits<allocator_type>;
+
+      // This is the state that is actually heap-allocated.
+      struct state {
+        template(typename... Args)                     //
+            (requires constructible_from<T, Args...>)  //
+            explicit state(
+                std::allocator_arg_t,
+                allocator_type allocator,
+                std::in_place_type_t<T>,
+                Args&&... args)
+          : object(static_cast<Args&&>(args)...)
+          , allocator(std::move(allocator)) {}
+
+        UNIFEX_NO_UNIQUE_ADDRESS T object;
+        UNIFEX_NO_UNIQUE_ADDRESS allocator_type allocator;
+      };
+
+      // This is the base-class object that holds the pointer to the
+      // heap-allocated state.
+      class base {
+      public:
+        template(typename... Args)                     //
+            (requires constructible_from<T, Args...>)  //
+            explicit base(
+                std::allocator_arg_t,
+                allocator_type allocator,
+                std::in_place_type_t<T>,
+                Args&&... args) {
+          state_ = allocator_traits::allocate(allocator, 1);
+
+          auto deallocateOnExit = scope_guard{[&]() noexcept {
+            allocator_traits::deallocate(allocator, state_, 1);
+          }};
+
+          allocator_traits::construct(
+              allocator,
+              state_,
+              std::allocator_arg,
+              allocator,
+              std::in_place_type<T>,
+              static_cast<Args&&>(args)...);
+
+          deallocateOnExit.release();
+        }
+
+        template(typename... Args)                   //
+            (requires                                //
+             constructible_from<T, Args...> AND      //
+             (!same_as<Allocator, allocator_type>))  //
+            explicit base(
+                std::allocator_arg_t,
+                Allocator allocator,
+                std::in_place_type_t<T>,
+                Args&&... args)
+          : base(
+                std::allocator_arg,
+                allocator_type(std::move(allocator)),
+                std::in_place_type<T>,
+                static_cast<Args&&>(args)...) {}
+
+        // Ideally this would only be instantiated if it's actually called.
+        //
+        // This should have a 'requires std::copy_constructible<T>' on the
+        // definition but we can't currently do that with the concepts emulation
+        // layer so we'll just enable the check when we actually have concepts.
+        base(const base& other)
+#if UNIFEX_CXX_CONCEPTS
+            requires copy_constructible<T>
+#endif
+          : base(
```

Comment:
Yes, even if you have a template function that would match the signature of a copy-constructor, this doesn't prevent the compiler for generating a default copy-constructor for you and calling that instead.

So we need to explicitly declare a copy-constructor here to prevent the compiler for generating a default one.